File: src/Python/1-100/39.CombinationSum.py
from copy import deepcopy

class Solution:
    def combinationSum(self, candidates: 'List[int]', target: 'int') -> 'List[List[int]]':
        ans = []
        r = []
        def recursive(r, ans, candidates, target, num):
            if target == 0:
                r.append(ans)
                return
            if target < candidates[0]:
                return
            for i in range(num, len(candidates)):
                if candidates[i] > target:
                    break
                l = deepcopy(ans)
                l.append(candidates[i])
                recursive(r, l, candidates, target - candidates[i], i)
        candidates.sort()
        recursive(r, ans, candidates, target, 0)
        return r

if __name__ == "__main__":
    solution = Solution()
    print(solution.combinationSum(candidates = [2,3,6,7], target = 7))
    print(solution.combinationSum(candidates = [2,3,5], target = 8))

File: src/Java/1-100/39.CombinationSum.java

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Queue;
import java.util.HashSet;
import java.util.LinkedList;

class Solution {
    List<List<Integer>> ans = new LinkedList<>();
    List<Integer> t = new LinkedList<>();

    void dfs(int i, int j, int[] v){
        if(j == 0){
            ans.add(new LinkedList<>(t));
            return;
        }
        if(i == v.length)
            return;    
        if(j >= v[i]){
            t.add(v[i]);
            dfs(i, j - v[i], v);
            t.remove(t.size() - 1);
        }
        dfs(i + 1, j, v);
        return;
    }

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        dfs(0, target, candidates);
        return ans;
    }
}